chiselFrontend/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -299,23 +299,58 @@ private[chisel3] object Builder {
   def warning(m: => String): Unit = if (dynamicContextVar.value.isDefined) errors.warning(m)
   def deprecated(m: => String, location: Option[String] = None): Unit =
     if (dynamicContextVar.value.isDefined) errors.deprecated(m, location)
-
-  /** Record an exception as an error, and throw it.
+  /** Given a string, record it as an error, generate an exception, and throw it.
+    *
+    * @param m exception message
+    * @param result result to be returned by this method
+    * @return result passed in by caller.
+    */
+  @throws(classOf[ChiselException])
+  def exception[T](m: => String, result: T): T = {
+    error(m)
+    result
+  }
+  /** Given a string, record it as an error, generate an exception, and throw it.
     *
     * @param m exception message
+    * @return does not return.
     */
   @throws(classOf[ChiselException])
-  def exception(m: => String): Unit = {
+  def exception(m: => String): Nothing = {
     error(m)
     throwException(m)
   }
+  /** Record an exception as an error, and return the result specified by the caller.
+    *
+    * @param e exception
+    * @param result result to be returned by this method.
+    * @return result passed in by caller.
+    */
+  def exception[T](e: Exception, result: T): T = {
+    error(e.getMessage)
+    result
+  }
+  /** Record an exception as an error, and throw it.
+    *
+    * @param e exception
+    * @return This method does not return.
+    */
+  @throws(classOf[Exception])
+  def exception(e: Exception): Nothing = {
```

Comment:
Should this maybe be `fatalException`? Is it a good idea to have the presence of an argument determine the difference between whether or not the error is recorded right away?

Also, should Chisel devs generally use the no-throw version of the throw version by default? I'd imagine you'd want to use the former as possible since otherwise it means that compilation terminates on the first exception.

Reply:
The basic idea is to make it easy to switch between **fatal** and **non-fatal** errors, assuming the rest of the code is in place to deal with the **non-fatal** condition. That may be an invalid assumption since in principle, additional code is required to make **non-fatal** work correctly.

I suspect Chisel devs will start with the **fatal** version of the error (as it is simpler to implement) before investing the additional time and effort into providing the additional support for the **non-fatal** condition.

This would be simplified if we had a general guide for implementing error diagnostics in Chisel, i.e. something like:
```scala
def chiselMethodWithErrorChecking(args: Data*): Data = {
  def validateArgs: Option[Seq[Error]]  = {
    val errors: ArrayBuffer[Error]
    // Each validate will update errors and return true or false depending on whether or not
    //  the remaining validation tests should be run.
    if (validateTest1 && validateTest2 && (validateTest3 || validateTest4)) {
      validateTestn
    }
    if (errors.isEmpty) {
      None
    } else {
      Some(errors.toSeq)
    }
  }
  validateArgs: match {
    case None =>
      someResult(args)
    case Some(errors: Seq[Error]) =>
      errors foreach Builder.exception(_.error, {} )
      ErrorData
  }
}
```
where `ErrorData` would be something like `DontCare` but would work for LHS (connect operations) as well.